esphome/components/sml/sml.cpp

```diff
@@ -0,0 +1,110 @@
+#include "sml.h"
+#include "esphome/core/log.h"
+#include "sml_parser.h"
+
+namespace esphome {
+namespace sml {
+
+static const char *const TAG = "sml";
+
+const char START_BYTES_DETECTED = 1;
+const char END_BYTES_DETECTED = 2;
+
+SmlListener::SmlListener(std::string server_id, std::string obis)
+    : server_id(std::move(server_id)), obis(std::move(obis)) {}
+
+char Sml::check_start_end_bytes_(char c) {
+  // fill sml_file with incoming bytes
+  for (int k = 1; k < 8; k++)
+    this->incoming_buffer_[k - 1] = this->incoming_buffer_[k];
```

Comment:
```suggestion
  std::rotate(incoming_buffer_, incoming_buffer_ + 1, incoming_buffer_ + 8);
```

Nit: could be more idiomatic C++ (might have to pull in `#include <algorithm>`)

Reply:
Done